### ISSUE TITLE
Allowing to pass a generic type to the compute function.

### ISSUE THREAD

**Issue by UliPrantz** [flutter/flutter#65213]
I tried to make a generic deserialization function for my data models in flutter which would use compute since I don't want huge json data to block my UI isolate. The problem is casting huge lists of data models also can be time consuming so I thought about a generic approach since I don't want to write an extra method for every data class I have. The problem is I couldn't manage to pass a type to the compute function. I tried it the following way: 

```dart
List<T> modelListFromJsonString<T>(String json) {
  // parse the json string and construct the models here 
  // decide which model to use based on the type T example:
  List<T> models = List();
  // do this for every model from the json String
  switch (T) {
    case T == User:
      models.add( User.fromJson(parsedJsonMap));
    case T == Post:
      models.add(Post.fromJson(parsedJsonMap));
  }
  return models;
}
```

This is a simplified version of the generic deserialization function I came up with. Which I tried to call the following ways:

```dart
compute(modelListFromJsonString, someJsonString); // meaningless since no type is passed T will always be dynamic
compute(modelListFromJsonString<User>, someJsonString); // not allowed syntax error
compute((String someJsonString) => modelListFromJsonString<User>(someJsonString) , someJsonString); // runtime error since compute must be called with a top level function closures are not allowed
```

Because of this I had to write an extra deserialization function for every data class just to use compute. An easy solution would be to give compute also a type parameter which it uses to call the given function callback. It wouldn't be a breaking change since if it isn't used, it just will be dynamic and nothing will happen. I thought about something like this:

```dart
compute<User>(modelListFromJsonString, someJsonString); 
// which would internally call modelListFromJsonString like this:
modelListFromJsonString<User>(someJsonString);
```

Also already discussed this problem in following [reddit thread](https://www.reddit.com/r/flutterhelp/comments/ikvl8d/how_to_use_compute_with_generic_types/) (subreddit r/flutterhelp)

If I missed some other easy solution please just point me in the right direction.

**Comment by TahaTesser** [comment]
Hi @UliPrantz 
This is related to Dart than Flutter itself 

Please open the issue in the dedicated [repository](https://github.com/dart-lang/sdk/issues).

Closing, as this isn't an issue with Flutter itself.  
If you disagree, please write in the comments, I will reopen it.
Thank you

**Comment by UliPrantz** [comment]
As far as I know the compute function is part of the package:flutter/foundation.dart which is a flutter package. Therefore the implementation is part of the flutter framework and flutter api and not part of the dart sdk.

I would say this makes this problem are flutter related one and not a dart one. I can't use the compute function in a pure dart project because it's part of flutter this should make this a flutter related problem.

**Comment by mraleph** [comment]
One problem here is that `SendPort` does not preserve type arguments of partially instantiated functions when they cross isolate boundary (https://github.com/dart-lang/sdk/issues/43343) which makes passing `modelListFromJsonString<User>` across the boundary impossible (even if you should be able to actually trigger partial instantiation by specifying `R` when you invoke `compute`, e.g. by writing `compute<String, User>(...)`).

As a work-around you just need to create a helper object which can carry type arguments over the boundary for you:

```dart
class DeserializeAction<T> {
  final String json;
  Deserialize(this.json);

  List<T> invoke() {
    return modelListFromJsonString<T>(json);
  }

  static dynamic _invoke(DeserializeAction a) => a.invoke();
}

T deserialize<T>(String json) => compute(DeserializeAction._invoke, DeserializeAction<T>(json));
```

**Comment by YahiaRagae** [comment]
Hi @mraleph, 
The `message` you are trying to pass to `compute` here (`DeserializeAction<T>(json)`) is an instance object which makes `compute` to throw that exception : `Illegal argument in isolate message : (object is a closure - Function '<anonymous closure>':.)`

**Comment by aam** [comment]
Fix for dart-lang/sdk#43343 made it to flutter master channel unblocking experiments with generic flutter `compute` mentioned above.

**Comment by mraleph** [comment]
I think we can close this now given that `compute<String, User>(modelListFromJsonString, someJsonString);` should now work as one would expect (at least once the change rolls into `master`).
